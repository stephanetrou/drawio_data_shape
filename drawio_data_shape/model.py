from dataclasses import dataclass

from drawio_data_shape.color import COLOR


@dataclass(frozen=True)
class Datasource:
    height: int
    width: int


@dataclass(frozen=True)
class IconDatasource(Datasource):
    shape_icon: str
    shape_icon_color: COLOR
    shape_group_icon: str = "mxgraph.aws4.resourceIcon"


@dataclass(frozen=True)
class ImageDatasource(Datasource):
    image: str


class AWSThemes:
    S3 = IconDatasource(20, 20, "mxgraph.aws4.s3", COLOR.AWS_STORAGE)
    RDS = IconDatasource(20, 20, "mxgraph.aws4.rds", COLOR.AWS_DATABASE)
    REDSHIFT = IconDatasource(20, 20, "mxgraph.aws4.redshift", COLOR.AWS_DATABASE)
    DYNAMODB = IconDatasource(20, 20, "mxgraph.aws4.dynamodb", COLOR.AWS_DATABASE)
    GLUE = IconDatasource(20, 20, "mxgraph.aws4.glue", COLOR.AWS_ANALYTICS)
    ATHENA = IconDatasource(20, 20, "mxgraph.aws4.athena", COLOR.AWS_ANALYTICS)


class OnPremTheme:
    ORACLE = IconDatasource(20, 20, "mxgraph.aws4.database", COLOR.ORACLE)


class AzureThemes:
    API = ImageDatasource(12, 19, "img/lib/azure2/integration/API_Management_Services.svg")
    POWER_BI = ImageDatasource(20, 15, "img/lib/azure2/analytics/Power_BI_Embedded.svg")
    CUBE_BLUE = ImageDatasource(20, 18, "img/lib/active_directory/database_cube.svg")
    BLOB = ImageDatasource(17, 20, "img/lib/mscae/BlobBlock.svg")
    SQL_DATA_WAREHOUSE = ImageDatasource(20, 20, "img/lib/azure2/databases/SQL_Data_Warehouses.svg")
    SQL_SERVER = ImageDatasource(20, 20, "img/lib/azure2/databases/SQL_Server.svg")


@dataclass(frozen=True)
class TemplateDetails:
    title: str
    shape_color: COLOR
    placeholder: str
    style: Datasource
    width: int = 210
    height: int = 70
    aspect: str = "fixed"


MX_GRAPH_MODELS = [
    # Sources
    TemplateDetails("source S3", COLOR.SOURCE, "Source S3", AWSThemes.S3),
    TemplateDetails("source rds", COLOR.SOURCE, "Source RDS", AWSThemes.RDS),
    TemplateDetails("source redshift", COLOR.SOURCE, "Source Redshift", AWSThemes.REDSHIFT),
    TemplateDetails("source dynamodb", COLOR.SOURCE, "Source DynamoDB", AWSThemes.DYNAMODB),
    TemplateDetails("source glue", COLOR.SOURCE, "Source Glue", AWSThemes.GLUE),
    TemplateDetails("source athena", COLOR.SOURCE, "Source Athena", AWSThemes.ATHENA),
    TemplateDetails("Source api", COLOR.SOURCE, "Source API", AzureThemes.API),
    TemplateDetails("source blob storage", COLOR.SOURCE, "Source Blob Storage", AzureThemes.BLOB),
    TemplateDetails(
        "source sql data warehouse", COLOR.SOURCE, "Source SQL Data Warehouse", AzureThemes.SQL_DATA_WAREHOUSE
    ),
    TemplateDetails("source sql server", COLOR.SOURCE, "Source SQL Server", AzureThemes.SQL_SERVER),
    TemplateDetails("source oracle", COLOR.SOURCE, "Source Oracle", OnPremTheme.ORACLE),
    # Models
    TemplateDetails("model S3", COLOR.MODEL, "Model S3", AWSThemes.S3),
    TemplateDetails("model rds", COLOR.MODEL, "Model RDS", AWSThemes.RDS),
    TemplateDetails("model redshift", COLOR.MODEL, "Model Redshift", AWSThemes.REDSHIFT),
    TemplateDetails("model dynamodb", COLOR.MODEL, "Model DynamoDB", AWSThemes.DYNAMODB),
    # Consumers
    TemplateDetails("consumer powerbi", COLOR.CONSUMER, "Consumer PowerBI", AzureThemes.POWER_BI),
    TemplateDetails("consumer cube blue", COLOR.CONSUMER, "Cube", AzureThemes.CUBE_BLUE),
    TemplateDetails("consumer api", COLOR.CONSUMER, "Consumer API", AzureThemes.API),
    TemplateDetails("consumer sql server", COLOR.CONSUMER, "Consumer SQL Server", AzureThemes.SQL_SERVER),
]


@dataclass(frozen=True)
class MedallionTemplate:
    title: str
    shape_color: COLOR
    width: int = 16
    height: int = 16
    aspect: str = "fixed"


MEDALLIONS = [
    MedallionTemplate("Bronze", COLOR.BRONZE),
    # MedallionTemplate("Dark Bronze", COLOR.DARK_BRONZE),
    MedallionTemplate("Silver", COLOR.SILVER),
    MedallionTemplate("Gold", COLOR.GOLD),
    # MedallionTemplate("Platinum", COLOR.PLATINUM),
    MedallionTemplate("Diamond", COLOR.DIAMOND),
    # MedallionTemplate("Light Diamond", COLOR.LIGHT_DIAMOND),
]
