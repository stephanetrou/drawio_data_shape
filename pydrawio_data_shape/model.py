from dataclasses import dataclass

from pydrawio_data_shape.color import COLOR


@dataclass(frozen=True)
class Style:
    height: int
    width: int


@dataclass(frozen=True)
class AWSStyle(Style):
    shape_icon: str
    shape_icon_color: COLOR
    shape_group_icon: str = "mxgraph.aws4.resourceIcon"


@dataclass(frozen=True)
class ImageStyle(Style):
    image: str


class AWSStyles:
    S3 = AWSStyle(20, 20, "mxgraph.aws4.s3", COLOR.AWS_STORAGE)
    RDS = AWSStyle(20, 20, "mxgraph.aws4.rds", COLOR.AWS_DATABASE)
    REDSHIFT = AWSStyle(20, 20, "mxgraph.aws4.redshift", COLOR.AWS_DATABASE)
    DYNAMODB = AWSStyle(20, 20, "mxgraph.aws4.dynamodb", COLOR.AWS_DATABASE)
    GLUE = AWSStyle(20, 20, "mxgraph.aws4.glue", COLOR.AWS_ANALYTICS)
    ATHENA = AWSStyle(20, 20, "mxgraph.aws4.athena", COLOR.AWS_ANALYTICS)


class OnPremStyles:
    ORACLE = AWSStyle(20, 20, "mxgraph.aws4.database", COLOR.ORACLE)
    ORACLE_MINI = AWSStyle(16, 20, "mxgraph.office.databases.database_mini_2", COLOR.ORACLE)


class AzureStyles:
    API = ImageStyle(15, 19, "img/lib/azure2/integration/API_Management_Services.svg")
    POWER_BI = ImageStyle(20, 15, "img/lib/azure2/analytics/Power_BI_Embedded.svg")
    CUBE_BLUE = AWSStyle(20, 18, "mxgraph.office.databases.database_cube", COLOR.CUBE_BLUE)
    BLOB = ImageStyle(17, 20, "img/lib/mscae/BlobBlock.svg")
    SQL_DATA_WAREHOUSE = ImageStyle(20, 20, "img/lib/azure2/databases/SQL_Data_Warehouses.svg")
    SQL_SERVER = ImageStyle(20, 20, "img/lib/azure2/databases/SQL_Server.svg")


@dataclass(frozen=True)
class MxGraphModel:
    title: str
    shape_color: COLOR
    placeholder: str
    style: Style
    width: int = 210
    height: int = 70
    aspect: str = "fixed"


MX_GRAPH_MODELS = [
    # Sources
    MxGraphModel("source S3", COLOR.SOURCE, "Source S3", AWSStyles.S3),
    MxGraphModel("source rds", COLOR.SOURCE, "Source RDS", AWSStyles.RDS),
    MxGraphModel("source redshift", COLOR.SOURCE, "Source Redshift", AWSStyles.REDSHIFT),
    MxGraphModel("source dynamodb", COLOR.SOURCE, "Source DynamoDB", AWSStyles.DYNAMODB),
    MxGraphModel("source glue", COLOR.SOURCE, "Source Glue", AWSStyles.GLUE),
    MxGraphModel("source athena", COLOR.SOURCE, "Source Athena", AWSStyles.ATHENA),
    MxGraphModel("Source api", COLOR.SOURCE, "Source API", AzureStyles.API),
    MxGraphModel("source blob storage", COLOR.SOURCE, "Source Blob Storage", AzureStyles.BLOB),
    MxGraphModel(
        "source sql data warehouse", COLOR.SOURCE, "Source SQL Data Warehouse", AzureStyles.SQL_DATA_WAREHOUSE
    ),
    MxGraphModel("source sql server", COLOR.SOURCE, "Source SQL Server", AzureStyles.SQL_SERVER),
    MxGraphModel("source oracle", COLOR.SOURCE, "Source Oracle", OnPremStyles.ORACLE),
    MxGraphModel("source oracle mini", COLOR.SOURCE, "Source Oracle", OnPremStyles.ORACLE_MINI),
    # Models
    MxGraphModel("model S3", COLOR.MODEL, "Model S3", AWSStyles.S3),
    MxGraphModel("model rds", COLOR.MODEL, "Model RDS", AWSStyles.RDS),
    MxGraphModel("model redshift", COLOR.MODEL, "Model Redshift", AWSStyles.REDSHIFT),
    MxGraphModel("model dynamodb", COLOR.MODEL, "Model DynamoDB", AWSStyles.DYNAMODB),
    # Consumers
    MxGraphModel("consumer powerbi", COLOR.CONSUMER, "Consumer PowerBI", AzureStyles.POWER_BI),
    MxGraphModel("consumer cube blue", COLOR.CONSUMER, "Cube", AzureStyles.CUBE_BLUE),
    MxGraphModel("consumer api", COLOR.CONSUMER, "Consumer API", AzureStyles.API),
    MxGraphModel("consumer sql server", COLOR.CONSUMER, "Consumer SQL Server", AzureStyles.SQL_SERVER),
]
