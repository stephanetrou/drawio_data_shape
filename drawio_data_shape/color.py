from enum import StrEnum


class COLOR(StrEnum):
    DEFAULT_COLOR = "#535C71"

    SOURCE = "#60A917"
    MODEL = "#1BA1E2"
    CONSUMER = "#6A00FF"

    AWS_STORAGE = "#7AA116"
    AWS_DATABASE = "#C925D1"
    AWS_ANALYTICS = "#8C4FFF"

    ORACLE = "#DA4026"
    CUBE_BLUE = "#2072B8"

    BRONZE = "#F0A30A"
    DARK_BRONZE = "#CD7F32"
    SILVER = "#C0C0C0"
    GOLD = "#FFD700"
    PLATINUM = "#E5E4E2"
    DIAMOND = "#70D1F4"
    LIGHT_DIAMOND = "#B9F2FF"
