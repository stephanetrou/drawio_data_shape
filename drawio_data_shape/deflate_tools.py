import base64
import zlib
from urllib.parse import quote, unquote


def decode(encoded: str) -> str:
    b64 = base64.b64decode(encoded)
    decompress_obj = zlib.decompressobj(-zlib.MAX_WBITS)
    uri_encoded = decompress_obj.decompress(b64)

    return unquote(uri_encoded)


def encode(decoded: str) -> str:
    quoted = quote(decoded)
    compress_obj = zlib.compressobj(method=zlib.DEFLATED, wbits=-zlib.MAX_WBITS)
    compressed = compress_obj.compress(quoted.encode())
    compressed += compress_obj.flush()
    return base64.b64encode(compressed).decode()
