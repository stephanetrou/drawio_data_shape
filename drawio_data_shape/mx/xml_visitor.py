from __future__ import annotations

from dataclasses import asdict
from typing import Any
from xml.etree import ElementTree as ET

from drawio_data_shape.mx.mxmodels import (
    MxCell,
    MxGeometry,
    MxGraphModel,
    MxObject,
    MxRectangle,
    UserObject,
    Visitor,
    has_default_value,
)


class XMLVisitor(Visitor):
    _hierarchy: list[ET.Element]

    def __init__(self):
        self._hierarchy = []

    def _append_hierarchical_element(self, element: ET.Element) -> None:
        """
        Append the element to the hierarchy and set it as the current element
        """
        if self._hierarchy:
            self._hierarchy[-1].append(element)

        self._hierarchy.append(element)

    def _pop_hierarchical_element(self) -> None:
        """
        Remove the current element from the hierarchy
        """
        if len(self._hierarchy) > 1:
            self._hierarchy.pop()

    @staticmethod
    def _create_element(key: str, value=None) -> ET.Element:
        """
        Create an XML element with the given key and value
        """
        element = ET.Element(key)

        if value is not None:
            element.text = str(value)

        return element

    @staticmethod
    def _add_attributes(element: ET.Element, visited: Any) -> None:
        for key, value in asdict(visited).items():
            if has_default_value(visited, key):
                continue

            if isinstance(value, bool):
                element.attrib[key] = "1" if value else "0"
            elif isinstance(value, str | int | float):
                element.attrib[key] = str(value)

    def visitMxGraphModel(self, mx_graph_model: MxGraphModel, **kwargs) -> None:
        xml_graph_model = XMLVisitor._create_element("mxGraphModel")
        self._append_hierarchical_element(xml_graph_model)

        xml_root = XMLVisitor._create_element("root")
        self._append_hierarchical_element(xml_root)

        for item in mx_graph_model.root:
            item.accept(self)

        self._pop_hierarchical_element()
        self._pop_hierarchical_element()

    def visitMxObject(self, mx_object: MxObject, **kwargs) -> None:
        xml_object = XMLVisitor._create_element("object")
        self._append_hierarchical_element(xml_object)

        XMLVisitor._add_attributes(xml_object, mx_object)
        if mx_object.cell is not None:
            mx_object.cell.accept(self)

        self._pop_hierarchical_element()

    def visitMxCell(self, mx_cell: MxCell, **kwargs) -> None:
        xml_cell = XMLVisitor._create_element("mxCell")

        self._append_hierarchical_element(xml_cell)

        XMLVisitor._add_attributes(xml_cell, mx_cell)

        if mx_cell.style is not None:
            styles = [f"{k}={'none' if v is None else v}" for k, v in mx_cell.style.items()]
            xml_cell.attrib["style"] = ";".join(styles)

        if mx_cell.geometry is not None:
            mx_cell.geometry.accept(self, _as="geometry")

        self._pop_hierarchical_element()

    def visitMxGeometry(self, mx_geometry: MxGeometry, **kwargs) -> None:
        xml_geometry = XMLVisitor._create_element("mxGeometry")
        self._append_hierarchical_element(xml_geometry)

        XMLVisitor._add_attributes(xml_geometry, mx_geometry)

        if "_as" in kwargs:
            xml_geometry.attrib["as"] = kwargs["_as"]

        if mx_geometry.alternateBounds is not None:
            mx_geometry.alternateBounds.accept(self, _as="alternateBounds")

        self._pop_hierarchical_element()

    def visitMxRectangle(self, mx_rectangle: MxRectangle, **kwargs) -> None:
        xml_rectangle = XMLVisitor._create_element("mxRectangle")
        self._append_hierarchical_element(xml_rectangle)

        XMLVisitor._add_attributes(xml_rectangle, mx_rectangle)
        if "_as" in kwargs:
            xml_rectangle.attrib["as"] = kwargs["_as"]

        self._pop_hierarchical_element()

    def visitUserModel(self, user_object: UserObject, **kwargs) -> None:
        xml_user_object = XMLVisitor._create_element("UserObject")
        self._append_hierarchical_element(xml_user_object)

        XMLVisitor._add_attributes(xml_user_object, user_object)
        if user_object.cell is not None:
            user_object.cell.accept(self)

        self._pop_hierarchical_element()

    def xml_str(self) -> str:
        return ET.tostring(self._hierarchy[0], encoding="unicode")
