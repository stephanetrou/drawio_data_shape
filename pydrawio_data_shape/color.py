from enum import StrEnum


class COLOR(StrEnum):
    SOURCE = "#60A917"
    MODEL = "#1BA1E2"
    CONSUMER = "#6A00FF"

    AWS_STORAGE = "#7AA116"
    AWS_DATABASE = "#C925D1"
    AWS_ANALYTICS = "#8C4FFF"
