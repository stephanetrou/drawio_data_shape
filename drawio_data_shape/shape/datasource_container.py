from drawio_data_shape.library_models import DatasourceTemplate, IconDatasource, ImageDatasource
from drawio_data_shape.models.shape import ShapeInformation, generate_encoded_shape
from drawio_data_shape.mx.mxmodels import MxCell, MxGeometry, MxGraphModel, MxObject, MxRectangle, UserObject
from drawio_data_shape.mx.mxstyle import StyleBuilder


class DatasourceContainer:
    def __init__(self):
        self.next_id = -1

    def _get_next_id(self):
        self.next_id += 1
        return self.next_id

    def build(self, template_detail: DatasourceTemplate) -> MxGraphModel:
        mx_graph_model = MxGraphModel()

        cell_0 = MxCell(self._get_next_id())
        mx_graph_model.append(cell_0)

        cell_1 = MxCell(self._get_next_id())
        cell_1.parent = cell_0.id
        mx_graph_model.append(cell_1)

        box = self._generate_box(template_detail)
        box.cell.parent = cell_1.id
        mx_graph_model.append(box)

        icons_container = self._generate_icons_container()
        icons_container.parent = box.id
        mx_graph_model.append(icons_container)

        link = self._generate_documentation()
        link.cell.parent = icons_container.id
        mx_graph_model.append(link)

        squad = self._generate_squad()
        squad.cell.parent = icons_container.id
        mx_graph_model.append(squad)

        git = self._generate_git_repository()
        git.cell.parent = icons_container.id
        mx_graph_model.append(git)

        label = self._generate_label(template_detail.placeholder)
        label.parent = box.id
        mx_graph_model.append(label)

        icon_datasource = self._generate_icon_datasource(template_detail.style)
        icon_datasource.parent = box.id
        mx_graph_model.append(icon_datasource)

        return mx_graph_model

    def _generate_icons_container(self) -> MxCell:
        geometry = MxGeometry(x=9, y=43.25, width=195, height=20)
        style = (
            StyleBuilder()
            .container("1")
            .collapsible("0")
            .movable("0")
            .connectable("0")
            .allow_arrows("0")
            .stroke_color("none")
            .fill_color("#FAFAFA")
            .expand("0")
            .resizable("0")
            .build()
        )

        cell = MxCell(id=self._get_next_id(), value="", style=style, geometry=geometry, vertex=True)

        return cell

    def _generate_box(self, template_detail: DatasourceTemplate) -> MxObject:
        shape_information = ShapeInformation("datasource", template_detail.height, template_detail.width)
        shape = generate_encoded_shape(shape_information)

        style = (
            StyleBuilder()
            .html("1")
            .aspect("fixed")
            .shadow("1")
            .container("1")
            .collapsible("0")
            .connectable("1")
            .fill_opacity("100")
            .fill_color(template_detail.shape_color)
            .shape(f"stencil({shape})")
            .build()
        )

        mx_object = MxObject(id=self._get_next_id(), label="", tooltip="")

        mx_object.cell = MxCell(
            style=style,
            vertex=True,
            geometry=MxGeometry(
                width=template_detail.width,
                height=template_detail.height,
                alternateBounds=MxRectangle(width=template_detail.width, height=template_detail.height),
            ),
        )

        return mx_object

    def _generate_documentation(self) -> UserObject:
        user_object = UserObject(
            id=self._get_next_id(), link="modify link", linkTarget="_blank", tooltip="Documentation", label=""
        )

        style = (
            StyleBuilder()
            .shape("image")
            .editable_css_rules(".*")
            .vertical_label_position("bottom")
            .label_background_color("default")
            .vertical_align("top")
            .aspect("fixed")
            .image_aspect("0")
            .image(
                "data:image/svg+xml,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9Ii0wLjUgLTAuNSAyNy45NDcgMjgiIGhlaWdodD0iMjgiIHdpZHRoPSIyNy45NDciPiYjeGE7CSAgPHN0eWxlIHR5cGU9InRleHQvY3NzIj4uc3QwIHsgc3RvcC1jb2xvcjogcmdiKDkxLCAxMTUsIDEzOSk7IH0gLnN0MSB7IHN0b3AtY29sb3I6IHJnYigzNCwgNTMsIDcyKTsgfSA8L3N0eWxlPiYjeGE7CTxkZWZzPiYjeGE7CQk8bGluZWFyR3JhZGllbnQgZ3JhZGllbnRVbml0cz0idXNlclNwYWNlT25Vc2UiIHkyPSIyNS4yNjkiIHgyPSIyMy40NSIgeTE9IjMuNDc0IiB4MT0iMi4wMzQiIGlkPSJBIj4mI3hhOwkJCTxzdG9wIGNsYXNzPSJzdDAiLz4mI3hhOwkJCTxzdG9wIGNsYXNzPSJzdDEiIG9mZnNldD0iMSIvPiYjeGE7CQk8L2xpbmVhckdyYWRpZW50PiYjeGE7CTwvZGVmcz4mI3hhOwk8cGF0aCBmaWxsPSJ1cmwoI0EpIiBkPSJNMTIuMTg5IDIwLjk5MmMuMTQzLS4xNDQuMzEzLS4yNTcuNTAyLS4zMzJzLjM5LS4xMTEuNTkyLS4xMDVjLjQzOCAwIC43NjcuMjAxLjk4NS42MDIuMjkuMjkyLjQzOC42MzkuNDM4IDEuMDRzLS4xNDcuNzQ3LS40MzcgMS4wNGwtMi4wODEgMi4wMjNjLS43MTMuNzE4LTEuNTcgMS4yNzctMi41MTUgMS42NGE2Ljg0IDYuODQgMCAwIDEtMi43OS42MDJjLS45NjItLjAwMy0xLjkxMi0uMjA4LTIuNzktLjYwMi0uODc1LS4zNjQtMS42NzYtLjkxLTIuNDA2LTEuNjQtLjcwOS0uNjk5LTEuMjY3LTEuNTM3LTEuNjQtMi40NmE3LjQ1IDcuNDUgMCAwIDEtLjU0OC0yLjc5YzAtLjk0OC4xODItMS44NzguNTQ4LTIuNzg5cy45MS0xLjczMyAxLjY0LTIuNDYxbDQuMTU2LTQuMTAyYy43My0uNzMgMS41NS0xLjI3NyAyLjQ2LTEuNjQxYTcuMjUgNy4yNSAwIDAgMSAyLjg0NC0uNmMuOTQ4IDAgMS44NzguMiAyLjc5LjYuOTEyLjM2NiAxLjcxNS45MTIgMi40MDYgMS42NDFhMS40MiAxLjQyIDAgMCAxIC40MzggMS4wNGMwIC40MDEtLjE0NS43NDctLjQzNyAxLjAzOGExLjQyIDEuNDIgMCAwIDEtMS4wMzkuNDM4IDEuNDIgMS40MiAwIDAgMS0xLjAzOC0uNDM3Yy0uODQtLjkxLTEuODYtMS4zNjUtMy4wNjItMS4zNjVzLTIuMjQ0LjQ1NS0zLjExOCAxLjM2NWwtNC4xNTMgNC4xMDJjLS45MS44NC0xLjM2NSAxLjg2LTEuMzY1IDMuMDYzcy40NTUgMi4yNjEgMS4zNjUgMy4xNzNjLjg0Ljg3NSAxLjg2IDEuMzEzIDMuMDYzIDEuMzEzczIuMjQ0LS40MzcgMy4xMTktMS4zMTJsMi4wNzQtMi4wODF6bTEzLjAxNy0xOS4yNWMuNzI4LjY5MyAxLjI3NiAxLjQ5NSAxLjY0IDIuNDA2YTYuODQgNi44NCAwIDAgMSAuNjAyIDIuNzljMCAuOTQ5LS4yMDEgMS44NzgtLjYwMiAyLjc5LS4zNjQuODc1LS45MSAxLjY3Ny0xLjY0IDIuNDA2bC00LjEwMiA0LjFjLS43NjcuNzY3LTEuNjA1IDEuMzMtMi41MTcgMS42OTgtLjkwNC4zNjMtMS44NjguNTQ4LTIuODQyLjU0NmE3LjQ1IDcuNDUgMCAwIDEtMi43OS0uNTQ4Yy0uOTAxLS4zODMtMS43MDQtLjk2Mi0yLjM1Mi0xLjY5NmExLjM2IDEuMzYgMCAwIDEtLjQzNy0uOTgzYzAtLjQwMi4xNDUtLjc0Ny40MzgtMS4wNGExLjM2IDEuMzYgMCAwIDEgLjk4NS0uNDM3IDEuNDIgMS40MiAwIDAgMSAxLjAzOC40MzggMy44NSAzLjg1IDAgMCAwIDEuNDIzLjk4NSA0LjM5IDQuMzkgMCAwIDAgMy4zMzYgMGMuNTQ4LS4yMTkgMS4wNC0uNTQ4IDEuNDc3LS45ODVsNC4xNTYtNC4xNTZjLjg3NS0uODM4IDEuMzEzLTEuODU5IDEuMzEzLTMuMDYzcy0uNDM3LTIuMjU5LTEuMzEyLTMuMTcxYy0uODQtLjg3NS0xLjg2LTEuMzEyLTMuMDYyLTEuMzEycy0yLjI2MS40MzgtMy4xNzMgMS4zMTNsLTIuMDc1IDIuMDc3YTEuNDIgMS40MiAwIDAgMS0xLjAzOS40MzggMS40MiAxLjQyIDAgMCAxLTEuMDQtLjQzNyAxLjQyIDEuNDIgMCAwIDEtLjQzNy0xLjA0IDEuNDIgMS40MiAwIDAgMSAuNDM4LTEuMDM4bDIuMDc5LTIuMDgxYy43MjgtLjcyOCAxLjU0OS0xLjI3NiAyLjQ2LTEuNjRhNy4yNSA3LjI1IDAgMCAxIDIuODQyLS42Yy45NDggMCAxLjg3OC4yMDEgMi43OS42MDIuOTEuMzY0IDEuNzEzLjkxIDIuNDA2IDEuNjR6Ii8+JiN4YTs8L3N2Zz4=;"
            )
            .editable("1")
            .movable("1")
            .resizable("0")
            .rotatable("1")
            .deletable("1")
            .locked("0")
            .connectable("0")
            .allow_arrows("0")
            .expand("0")
            .meta_edit("0")
            .container("0")
            .build()
        )

        cell = MxCell(
            style=style,
            vertex=True,
            geometry=MxGeometry(x=7, y=5, width=10, height=10),
        )

        user_object.cell = cell

        return user_object

    def _generate_squad(self) -> UserObject:
        user_object = UserObject(id=self._get_next_id(), tooltip="Squad", label="")

        style = (
            StyleBuilder()
            .fill_color("#535C71")
            .stroke_color(None)
            .pointer_events("1")
            .align("center")
            .aspect("fixed")
            .connectable("0")
            .container("0")
            .allow_arrows("0")
            .resizable("0")
            .shape("mxgraph.cisco_safe.people_places_things_icons.icon9")
            .build()
        )

        cell = MxCell(
            style=style,
            vertex=True,
            geometry=MxGeometry(x=22.46, y=3.25, width=15, height=13.5),
        )

        user_object.cell = cell

        return user_object

    def _generate_git_repository(self) -> UserObject:
        user_object = UserObject(id=self._get_next_id(), link="Git link", linkTarget="_blank", tooltip="Git", label="")

        style = (
            StyleBuilder()
            .fill_color("#535C71")
            .stroke_color(None)
            .pointer_events("1")
            .align("center")
            .aspect("fixed")
            .connectable("0")
            .container("0")
            .allow_arrows("0")
            .resizable("0")
            .shape("mxgraph.azure.git_repository")
            .build()
        )

        cell = MxCell(
            style=style,
            vertex=True,
            geometry=MxGeometry(x=42, y=3.25, width=15, height=15),
        )

        user_object.cell = cell

        return user_object

    def _generate_label(self, placeholder) -> MxCell:
        style = (
            StyleBuilder()
            .html("1")
            .align("left")
            .vertical_align("middle")
            .white_space("wrap")
            .rounded("0")
            .font_style("0")
            .font_family("Tahoma")
            .font_size("12")
            .font_color("#535C71")
            .stroke_color("none")
            .movable("0")
            .connectable("0")
            .allow_arrows("0")
            .expand("0")
            .resizable("0")
            .build()
        )

        geometry = MxGeometry(x=40, y=6, width=164, height=20)

        cell = MxCell(id=self._get_next_id(), value=placeholder, style=style, geometry=geometry, vertex=True)
        return cell

    def _generate_icon_datasource(self, style_icon) -> MxCell:
        style = self._generate_style_icon(style_icon)
        geometry = MxGeometry(x=15, y=5, width=style_icon.width, height=style_icon.height)

        cell = MxCell(id=self._get_next_id(), value="", style=style.build(), geometry=geometry, vertex=True)
        return cell

    def _generate_style_icon(self, style_icon) -> StyleBuilder:
        if isinstance(style_icon, IconDatasource):
            return (
                StyleBuilder()
                .align("center")
                .allow_arrows("0")
                .aspect("fixed")
                .connectable("0")
                .expand("0")
                .fill_color(style_icon.shape_icon_color)
                .font_color("#232F3E")
                .font_size("12")
                .font_style("0")
                .html("1")
                .outline_connect("0")
                .recursive_resize("0")
                .res_icon(style_icon.shape_icon)
                .resizable("0")
                .shape(style_icon.shape_group_icon)
                .sketch("0")
                .stroke_color("#FFFFFF")
                .vertical_align("top")
                .vertical_label_position("bottom")
            )
        elif isinstance(style_icon, ImageDatasource):
            return (
                StyleBuilder()
                .align("center")
                .image_align("center")
                .allow_arrows("0")
                .aspect("fixed")
                .connectable("0")
                .expand("0")
                .font_size("12")
                .font_style("0")
                .html("1")
                .image(style_icon.image)
                .stroke_color("none")
                .outline_connect("0")
                .recursive_resize("0")
                .resizable("0")
            )
