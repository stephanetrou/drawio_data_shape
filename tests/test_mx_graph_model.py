from approvaltests import verify_xml

from drawio_data_shape.color import COLOR
from drawio_data_shape.datasource_container import DatasourceContainer
from drawio_data_shape.model import AWSThemes, TemplateDetails
from drawio_data_shape.mx.models import MxCell, MxGeometry, MxGraphModel, MxRectangle, UserObject
from drawio_data_shape.mx.style import StyleBuilder
from drawio_data_shape.mx.xml_visitor import XMLVisitor


def test_mx_cell_empty():
    visitor = XMLVisitor()
    visitor.visitMxCell(MxCell(0))

    pretty_xml = visitor.xml_str()
    verify_xml(pretty_xml)


def test_mx_cell_parent():
    visitor = XMLVisitor()
    visitor.visitMxCell(MxCell(1, parent=0))

    pretty_xml = visitor.xml_str()
    verify_xml(pretty_xml)


def test_mx_geometry():
    visitor = XMLVisitor()
    visitor.visitMxGeometry(MxGeometry(height=12, width=15))

    pretty_xml = visitor.xml_str()
    verify_xml(pretty_xml)


def test_mx_rectangle():
    visitor = XMLVisitor()
    visitor.visitMxRectangle(MxRectangle(x=10, y=11, width=12, height=13))

    pretty_xml = visitor.xml_str()
    verify_xml(pretty_xml)


def test_mx_cell_with_geometry():
    visitor = XMLVisitor()
    visitor.visitMxCell(MxCell(2, geometry=MxGeometry(height=12, width=15)))

    pretty_xml = visitor.xml_str()
    verify_xml(pretty_xml)


def test_mx_cell_with_geometry_and_alternate_bounds():
    visitor = XMLVisitor()
    visitor.visitMxCell(MxCell(2, geometry=MxGeometry(height=12, width=15, alternateBounds=MxRectangle(x=10, y=11))))

    pretty_xml = visitor.xml_str()
    verify_xml(pretty_xml)


def test_user_object():
    visitor = XMLVisitor()
    visitor.visitUserModel(UserObject(id=1, label="Hello"))

    pretty_xml = visitor.xml_str()
    verify_xml(pretty_xml)


def test_user_object_with_mx_cell():
    visitor = XMLVisitor()
    visitor.visitUserModel(UserObject(id=1, label="Hello", cell=MxCell(2)))

    pretty_xml = visitor.xml_str()
    verify_xml(pretty_xml)


def test_simple_shape():
    graph_model = MxGraphModel()

    graph_model.append(MxCell(0))

    graph_model.append(MxCell(1, parent=0))

    style = (
        StyleBuilder()
        .rounded("1")
        .white_space("wrap")
        .html("1")
        .font_family("Verdana")
        .font_size("9")
        .stroke_color(None)
        .fill_color("#535C71")
        .font_color("#FFFFFF")
        .build()
    )

    geometry = MxGeometry(height=12, width=26)

    graph_model.append(MxCell(2, value="Hi !!", geometry=geometry, style=style, vertex=True, parent=1))

    xml_visitor = XMLVisitor()
    xml_visitor.visitMxGraphModel(graph_model)
    pretty_xml = xml_visitor.xml_str()

    verify_xml(pretty_xml)


def test_datasource_shape():
    template_detail = TemplateDetails("source S3", COLOR.SOURCE, "Source S3", AWSThemes.S3)

    container = DatasourceContainer()
    graph_model = container.build(template_detail)

    xml_visitor = XMLVisitor()
    xml_visitor.visitMxGraphModel(graph_model)
    pretty_xml = xml_visitor.xml_str()

    verify_xml(pretty_xml)
