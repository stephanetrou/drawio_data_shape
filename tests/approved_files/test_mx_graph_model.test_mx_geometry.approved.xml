<?xml version="1.0" ?>
<mxGeometry width="15" height="12"/>
